### ISSUE TITLE
Retinanet: differences between __call__ and __predict__

### ISSUE THREAD

**Issue by Dapid** [keras-team/keras-cv#2188]
### Current Behavior:
According to the Keras documentation, the `__call__` and `predict` should yield identical results -- at least for small batches. That isn't the case for RetinaNet, where we get an absurdly large amount of boxes.

For a trained model, the boxes in the raw output don't make much sense, running Non Max Supression from OpenCV doesn't help either (we go from 57k boxes to about 6k, when the `predict` method correctly identifies 3).

### Why is this important?

I need to export the model to TFLite, and that runs just `__call__`, not the post-processing step.

### Steps To Reproduce:

https://colab.research.google.com/gist/Dapid/42f2256d3c0f51583aa8447dde7f7090/untitled1.ipynb

https://gist.github.com/Dapid/44523d82874e7f137fa30df733dd23f4

Note that the trained model actually `predict`s correct boxes.

### Version:
- 0.6.4, 0.7.1

**Comment by innat-asj** [comment]
@Dapid 
https://github.com/keras-team/keras-cv/issues/1902

**Comment by Dapid** [comment]
Thanks.

I see it is intended, but the behaviour directly contradicting the documentation is bad, in my opinion.

I can contribute rewording the docstring, but it still leaves his top export to tflite. Simply applying NMS from OpenCV isn't enough, I suspect because I am not doing some steps correctly, and I haven't been able to trace down the exact code, there are so many layers of redirection.

**Comment by tirthasheshpatel** [comment]
> I can contribute rewording the docstrin,

Hi @Dapid, we would really appreciate documenting this in KerasCV. I think this is the right way to go for the time being (at least until the issue is fixed), a bit of explaination below.

> Simply applying NMS from OpenCV isn't enough, I suspect because I am not doing some steps correctly, and I haven't been able to trace down the exact code, there are so many layers of redirection.

Here's the exact code that does decoding + NMS for RetinaNet:

https://github.com/keras-team/keras-cv/blob/37ffac047e9d700235801cb8820982887a114e28/keras_cv/models/object_detection/retinanet/retinanet.py#L289-L320

You should be able to replicate the behaviour of the `predict` method by first calling `__call__` and then calling `model.decode_predictions` on the results. Let me know if doing this resolves your issue. Here's how KerasCV did it:

https://github.com/keras-team/keras-cv/blob/37ffac047e9d700235801cb8820982887a114e28/keras_cv/models/object_detection/predict_utils.py#L33-L58

Now, to get to your question; why are `__call__` and `predict` different: KerasCV used to have a custom NMS op which didn't work on the TPU, so, `RetinaNet` and `YOLOv8` avoided running the `decode_predictions` part when `__call__` was called; this enabled training the model on TPU. 

This doesn't apply anymore since the new [`NonMaximalSuppression` layer](https://github.com/keras-team/keras-cv/blob/37ffac047e9d700235801cb8820982887a114e28/keras_cv/layers/object_detection/non_max_suppression.py#L29) works in a backend-agnostic manner. YOLOv8 was updated a few months ago to use this. We can also update the RetinaNet model and remove this special behaviour now.

**Comment by Dapid** [comment]
Thank you, that is very helpful.

I think normalising the behaviour is the right approach. How much effort would it take?

**Comment by tirthasheshpatel** [comment]
I just realized the RetinaNet model was also updated along with YOLOv8:

https://github.com/keras-team/keras-cv/blob/37ffac047e9d700235801cb8820982887a114e28/keras_cv/models/object_detection/retinanet/retinanet.py#L247-L252

So, no fix required here.

I was suggesting adding docs for how to make `__call__` equivalent to `predict`. @Dapid Are you interesting in contributing this?

We still should avoid running `decode_prediction` if the model is called directly since all the predictions are used to train the model. Only when infering, `decode_prediction` needs to run. I hope this clarifies everything.

(By fixing the "special behavior" in my previous comment, I just meant removing the use of the custom op but that was done when porting the model to Keras 3.)
